src/mainwin.cc

```diff
@@ -2013,6 +2017,166 @@ void MainWindow::csgRender()
 	compileEnded();
 }
 
+
+void MainWindow::action3DPrint()
+{
+	//Keeps track of how many times we've exported and tries to create slightly unique filenames.
+	//Not mission critical, since non-unique file names are fine for the API, just harder to 
+	//differentiate between in customer support later.
+	static unsigned int printCounter=0;
+	
+	setCurrentOutput();
+	PRINT("3D Printing...");
+	
+	unsigned int dim = 3;
+	
+	QUrl partUrl;
+	
+	//Where we hold our temporary stl file name:
+	char tempStlFileName[L_tmpnam];
+	
+	setCurrentOutput();
+
+	//Make sure we can export:
+	if (! canExport(dim))
+	{
+		PRINT("Cannot 3D Print due to errors.");
+		return;
+    }
+	
+	//Create a temporary file name valid on all systems:
+	QTemporaryFile exportFile;
+	//Open the file so we can get the name:
+	if ( ! exportFile.open())
+	{
+		PRINT("Could not open temporary file.");
+		return;
+	}
+	QString exportFilename=exportFile.fileName();
+	
+	//Render the stl to a temporary file:
+	exportFileByName(this->root_geom, FileFormat::STL,
+		exportFilename.toLocal8Bit().constData(),
+		exportFilename.toUtf8());
+	
+	//Create a name that the order process will use to refer to the file. Base it off of the
+	//project name
+	QString userFacingName="unsaved.stl";
+	if (!this->fileName.isEmpty()) {
+		QString fileBasename = QFileInfo(this->fileName).baseName();
+		userFacingName = fileBasename.append("_").append(QString::number(printCounter++)).append(".stl");
+	}
+	
+	//Upload the file to the 3D Printing server and get the corresponding url to see it.
+	//The result is put in partUrl.
+	bool uploadWorked=uploadStlAndGetPartUrl(exportFilename, userFacingName, partUrl);
+	setCurrentOutput();
+
+	//Check the result:
+	if (! uploadWorked)
+	{
+		PRINT("An error occured while contacting the 3D Print API.");
```

Comment:
Could we provide more information about what kind of error? e.g. HTTP status code + message?
A cleaner way of doing that could be to just throw an exception with this info instead of passing writable arguments.

Reply:
Oh.  I do provide information about the status code - that's in the function itself.  I agree this could be done in a more explicit way.  I'll look into how to do that.